lib/stateTransition/AbstractStateTransition.js

```diff
@@ -49,6 +77,148 @@ class AbstractStateTransition {
   hash() {
     return hash(this.serialize()).toString('hex');
   }
+
+  /**
+   *  Returns signature
+   *
+   * @return {string|null}
+   */
+  getSignature() {
+    return this.signature;
+  }
+
+  /**
+   * Returns public key id
+   *
+   * @returns {number|null}
+   */
+  getSignaturePublicKeyId() {
+    return this.signaturePublicKeyId;
+  }
+
+  /**
+   * Sign data and check identityPublicKey
+   *
+   * @param {IdentityPublicKey} identityPublicKey
+   * @param {string|Buffer|Uint8Array|PrivateKey} privateKey string must be hex or base58
+   * @return {AbstractStateTransition}
+   */
+  sign(identityPublicKey, privateKey) {
+    const data = this.serialize({ skipSignature: true });
+    let privateKeyModel;
+    let pubKeyBase;
+
+    switch (identityPublicKey.getType()) {
+      case IdentityPublicKey.TYPES.ECDSA_SECP256K1:
+        privateKeyModel = new PrivateKey(privateKey);
+
+        pubKeyBase = new PublicKey({
+          ...privateKeyModel.toPublicKey().toObject(),
+          compressed: true,
+        })
+          .toBuffer()
+          .toString('base64');
+
+        if (pubKeyBase !== identityPublicKey.getData()) {
+          throw new InvalidSignaturePublicKeyError(identityPublicKey.getData());
+        }
+
+        this.signature = signData(data, privateKeyModel).toString('base64');
+
+        break;
+      default:
+        throw new InvalidSignatureTypeError(identityPublicKey.getType());
+    }
+
+    this.signaturePublicKeyId = identityPublicKey.getId();
+
+    return this;
+  }
+
+  /**
+   * Sign data with private key
+   * @param {string|Buffer|Uint8Array|PrivateKey} privateKey string must be hex or base58
+   * @return {AbstractStateTransition}
+   */
+  signByPrivateKey(privateKey) {
+    const data = this.serialize({ skipSignature: true });
+    const privateKeyModel = new PrivateKey(privateKey);
+
+    this.signature = signData(data, privateKeyModel).toString('base64');
+
+    return this;
+  }
+
+  /**
+   * Verify signature
+   *
+   * @param {IdentityPublicKey} publicKey
+   * @return {boolean}
+   */
+  verifySignature(publicKey) {
+    const signature = this.getSignature();
+    if (!signature || publicKey.getId() !== this.signaturePublicKeyId) {
+      return false;
```

Comment:
In order to reduce ambiguity. `false` represents too much. I would throw errors here if public key ID mismatches or signature is absent.

Reply:
So it appears we don't need to do it here, as it already validate in the ST validation function

Reply:
I would check it anyways coz people can call it manually before signing and in this case, it would be nice to throw an error.